src/librustc_typeck/coherence/orphan.rs

```diff
@@ -92,6 +93,40 @@ impl<'cx, 'tcx,'v> visit::Visitor<'v> for OrphanChecker<'cx, 'tcx> {
                         }
                     }
                 }
+
+                // Impls of a defaulted trait face additional rules.
+                debug!("trait_ref={} trait_def_id={} trait_has_default_impl={}",
+                       trait_ref.repr(self.tcx),
+                       trait_def_id.repr(self.tcx),
+                       ty::trait_has_default_impl(self.tcx, trait_def_id));
+                if
+                    ty::trait_has_default_impl(self.tcx, trait_def_id) &&
+                    trait_def_id.krate != ast::LOCAL_CRATE
+                {
+                    let self_ty = trait_ref.self_ty();
+                    match self_ty.sty {
+                        ty::ty_struct(self_def_id, _) | ty::ty_enum(self_def_id, _) => {
+                            // The orphan check often rules this out,
+                            // but not always. For example, the orphan
+                            // check would accept `impl Send for
+                            // Box<SomethingLocal>`, but we want to
+                            // forbid that.
+                            if self_def_id.krate != ast::LOCAL_CRATE {
+                                self.tcx.sess.span_err(
+                                    item.span,
+                                    "cross-crate traits with a default impl \
+                                     can only be implemented for a struct/enum type \
```

Comment:
nit: can we make these 2 error messages consistent `struct/enum` vs `struct or enum`